core/src/main/scala/kafka/log/LogLoader.scala

```diff
@@ -368,12 +368,40 @@ object LogLoader extends Logging {
     for (swapFile <- swapFiles) {
       val logFile = new File(CoreUtils.replaceSuffix(swapFile.getPath, Log.SwapFileSuffix, ""))
       val baseOffset = Log.offsetFromFile(logFile)
+
+       def maybeCompleteInterruptedSwap(fn: (File, Long, String) => File): Boolean = {
+        val swapOffsetIndexFile = fn(swapFile.getParentFile, baseOffset, Log.SwapFileSuffix)
+        if (swapOffsetIndexFile.exists())
+          true
+        else {
+          val cleanedOffsetIndexFile = Log.offsetIndexFile(swapFile.getParentFile, baseOffset, Log.CleanedFileSuffix)
+          if (cleanedOffsetIndexFile.exists() && cleanedOffsetIndexFile.renameTo(swapOffsetIndexFile)) true
+          else false
+        }
+      }
+
+      // Check whether swap index files exist: if not, the cleaned files must exist due to the
+      // existence of swap log file. Therefore, we rename the cleaned files to swap files and continue.
+      val recoverable = maybeCompleteInterruptedSwap(Log.offsetIndexFile) &&
+        maybeCompleteInterruptedSwap(Log.timeIndexFile) &&
+        maybeCompleteInterruptedSwap(Log.transactionIndexFile)
       val swapSegment = LogSegment.open(swapFile.getParentFile,
         baseOffset = baseOffset,
         params.config,
         time = params.time,
         fileSuffix = Log.SwapFileSuffix)
-      info(s"${params.logIdentifier}Found log file ${swapFile.getPath} from interrupted swap operation, repairing.")
+      if (recoverable) {
+        try {
+          swapSegment.sanityCheck(true)
+          info(s"Found log file ${swapFile.getPath} from interrupted swap operation, which is recoverable from ${Log.CleanedFileSuffix} files.")
+          swapSegment.changeFileSuffixes(Log.SwapFileSuffix, "")
+          return
```

Comment:
`return` might be wrong. Didn't realize it was within a `for` loop. Maybe `continue` instead.

Reply:
Would be good to avoid `return` or `continue` and instead make the call to `recoverSegment` conditional so that the code is easier to read.